### PR TITLE
PHP File Upload improvements

### PR DESCRIPTION
**Pull request by lohanidamodar**

### CODE REVIEW

File: templates/php/src/Services/Service.php.twig [hunk 1]
```diff
@@ -58,43 +59,74 @@ class {{ service.name | caseUcfirst }} extends Service
...
-            fseek($handle, $start);
-            $params['{{ parameter.name }}'] = new \CURLFile('data://' . $mimeType . ';base64,' . base64_encode(@fread($handle, Client::CHUNK_SIZE)), $mimeType, $postedName);
+            $chunk = '';
+            if(!empty($handle)) {
+                fseek($handle, $start);
+                $chunk = @fread($handle, Client::CHUNK_SIZE);
+            } else {
+                $chunk = substr($file->getData(), $start, Client::CHUNK_SIZE);
+            }
+            $params['{{ parameter.name }}'] = new \CURLFile('data://' . $mimeType . ';base64,' . base64_encode($chunk), $mimeType, $postedName);
```
Comment (abnegate) [inline templates/php/src/Services/Service.php.twig:129]: Is there any way we can avoid base64 encoding each chunk?

File: tests/languages/php/test.php [hunk 2]
```diff
@@ -59,10 +61,10 @@
 $response = $general->redirect();
 echo "{$response['result']}\n";
 
-$response = $general->upload('string', 123, ['string in array'], __DIR__ . '/../../resources/file.png');
+$response = $general->upload('string', 123, ['string in array'], InputFile::withPath(__DIR__ .'/../../resources/file.png'));
 echo "{$response['result']}\n";
 
-$response = $general->upload('string', 123, ['string in array'], __DIR__ . '/../../resources/large_file.mp4');
+$response = $general->upload('string', 123, ['string in array'], InputFile::withPath(__DIR__ .'/../../resources/large_file.mp4'));
 echo "{$response['result']}\n";
```
Comment (Meldiron) [inline tests/languages/php/test.php:69]: Can we add `withData` test?

File: templates/php/src/Services/Service.php.twig [hunk 3]
```diff
@@ -58,43 +59,74 @@ class {{ service.name | caseUcfirst }} extends Service
...
 {% if parameter.isUploadID %}
-            if(${{ parameter.name | caseCamel | escapeKeyword }} != 'unique()') {
-                try {
-                    $response = $this->client->call(Client::METHOD_GET, new URL($path . '/' . {{ parameter.name }}), headers);
-                    $counter = $response['chunksUploaded'] ?? 0;
-                } catch(\Exception $e) {
-                }
+        if(${{ parameter.name | caseCamel | escapeKeyword }} != 'unique()') {
+            try {
+                $response = $this->client->call(Client::METHOD_GET, $path . '/' . ${{ parameter.name }},);
```
Comment (Meldiron) [inline templates/php/src/Services/Service.php.twig:105]: You removed headers here, and kept comma. Looks like syntax error, also unwanted change probably.

File: templates/php/src/InputFile.php.twig [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    public static function withPath(string $path, ?string $mimeType = null, ?string $filename = null)
+    {
+        $instance = new InputFile();
+        $instance->path = $path;
+        $instance->mimeType = $mimeType;
+        $instance->filename = $filename;
+        return $instance;
+    } 
+
+    public static function withData(string $data, ?string $mimeType = null, ?string $filename = null)
```
Comment (Meldiron) [inline templates/php/src/InputFile.php.twig:41]: Can we come up with a better name? Data sounds too genetic. For example, in Node I called it `plaintext`, not sure if it's the same concept here.

File: templates/php/src/Services/Service.php.twig (hunk 1 above)
Comment (lohanidamodar) [reply templates/php/src/Services/Service.php.twig:129]: We can drop later when we only support PHP 8.1 and above

File: templates/php/src/InputFile.php.twig (hunk 4 above)
Comment (lohanidamodar) [reply templates/php/src/InputFile.php.twig:41]: I've seen data been common for file data as string in PHP?

Comment (Meldiron) [reply templates/php/src/InputFile.php.twig:42]: Let's keep it if it's common in PHP.

**Review by Meldiron — APPROVED** [review]
LGTM

**Pull request merged**